File: app/src/main/java/com/tevinjeffrey/rutgersct/customviews/CircleView.java
package com.tevinjeffrey.rutgersct.customviews;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Canvas;
import android.graphics.Color;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Typeface;
import android.os.Parcelable;
import android.text.TextPaint;
import android.util.AttributeSet;
import android.view.View;

import com.tevinjeffrey.rutgersct.R;

import icepick.Icepick;
import icepick.Icicle;

public class CircleView extends View {

    final int DEFAULT_TITLE_COLOR = Color.CYAN;

    final String DEFAULT_TITLE = "Title";

    final boolean DEFAULT_SHOW_TITLE = true;

    final float DEFAULT_TITLE_SIZE = 25f;

    final int DEFAULT_BACKGROUND_COLOR = Color.WHITE;

    final int DEFAULT_VIEW_SIZE = 96;

    @Icicle
    int mTitleColor = DEFAULT_TITLE_COLOR;

    @Icicle
    int mBackgroundColor = DEFAULT_BACKGROUND_COLOR;

    @Icicle
    String mTitleText = DEFAULT_TITLE;

    @Icicle
    float mTitleSize = DEFAULT_TITLE_SIZE;

    @Icicle
    boolean mShowTitle = DEFAULT_SHOW_TITLE;

    TextPaint mTitleTextPaint;

    Paint mBackgroundPaint;

    @Icicle
    RectF mInnerRectF;

    @Icicle
    int mViewSize;

    public CircleView(Context context) {
        super(context);
        init(null, 0);
    }

    public CircleView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(attrs, 0);
    }

    public CircleView(Context context, AttributeSet attrs, int defStyle) {
        super(context, attrs, defStyle);
        init(attrs, defStyle);
    }

    private void init(AttributeSet attrs, int defStyle) {
        final TypedArray a = getContext().obtainStyledAttributes(
                attrs, R.styleable.CircleView, defStyle, 0);

        if(a.hasValue(R.styleable.CircleView_titleText)){
            mTitleText = a.getString(R.styleable.CircleView_titleText);
        }

        mTitleColor = a.getColor(R.styleable.CircleView_titleColor,DEFAULT_TITLE_COLOR);
        mBackgroundColor = a.getColor(R.styleable.CircleView_backgroundColorValue,DEFAULT_BACKGROUND_COLOR);

        mTitleSize = a.getDimension(R.styleable.CircleView_titleSize,DEFAULT_TITLE_SIZE);


        a.recycle();

        mTitleTextPaint = new TextPaint();
        mBackgroundPaint = new Paint();
        mInnerRectF = new RectF();

        //Title TextPaint
        mTitleTextPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mTitleTextPaint.setAntiAlias(true);
        mTitleTextPaint.setTypeface(Typeface.defaultFromStyle(Typeface.BOLD));
        mTitleTextPaint.setTextAlign(Paint.Align.CENTER);
        mTitleTextPaint.setLinearText(true);
        mTitleTextPaint.setColor(mTitleColor);
        mTitleTextPaint.setTextSize(mTitleSize);


        //Background Paint
        mBackgroundPaint.setFlags(Paint.ANTI_ALIAS_FLAG);
        mBackgroundPaint.setStyle(Paint.Style.FILL_AND_STROKE);
        mBackgroundPaint.setColor(mBackgroundColor);

    }


    private void invalidatePaints(){
        mBackgroundPaint.setColor(mBackgroundColor);
        invalidate();
    }

    @Override
    protected void onMeasure(int widthMeasureSpec, int heightMeasureSpec) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec);

        int width = resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec);
        int height = resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec);
        mViewSize = Math.min(width, height);

        setMeasuredDimension(width, height);
    }

    @Override
    protected void onDraw(Canvas canvas) {
        mInnerRectF.set(0, 0, mViewSize, mViewSize);

        final float centerX = mInnerRectF.centerX();
        final float centerY = mInnerRectF.centerY();

        canvas.drawCircle(centerX, centerY, mViewSize/2 ,mBackgroundPaint);

        final int xPos = (int) centerX;
        final int yPos = (int) (centerY - (mTitleTextPaint.descent() + mTitleTextPaint.ascent()) / 2);

        if (mShowTitle) {
            canvas.drawText(mTitleText,
                    xPos,
                    yPos,
                    mTitleTextPaint);
        }


    }

    @Override
    public boolean hasOverlappingRendering() {
        return false;
    }

    /**
     * Sets whether the view's title string will be shown.
     * @param flag The boolean value.
     */
    public void setShowTitle(boolean flag){
        this.mShowTitle = flag;
        invalidate();
    }

    /**
     * Sets the view's title string attribute value.
     * @param title The example string attribute value to use.
     */
    public void setTitleText(String title) {
        mTitleText = title;
        invalidate();
    }

    public String getTitleText() {
       return mTitleText;
    }

    /**
     * Sets the view's background color attribute value.
     * @param backgroundColor The background color attribute value to use.
     */
    public void setBackgroundColor(int backgroundColor) {
        mBackgroundColor = backgroundColor;
        invalidatePaints();
    }

    public int getBackgroundColor() {
        return mBackgroundColor;
    }

    @Override
    protected Parcelable onSaveInstanceState() {
        return Icepick.saveInstanceState(this, super.onSaveInstanceState());
    }

    @Override
    protected void onRestoreInstanceState(Parcelable state) {
        super.onRestoreInstanceState(Icepick.restoreInstanceState(this, state));
    }
}

File: app/src/main/java/com/tevinjeffrey/rutgersct/ui/utils/CircleView.kt
package com.tevinjeffrey.rutgersct.ui.utils

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.RectF
import android.graphics.Typeface
import android.os.Parcelable
import android.text.TextPaint
import android.util.AttributeSet
import android.view.View

import com.tevinjeffrey.rutgersct.R

import icepick.Icepick
import icepick.State

class CircleView
@JvmOverloads constructor(context: Context, attrs: AttributeSet, defStyle: Int = 0)
  : View(context, attrs, defStyle) {

  init {
    init(attrs, defStyle)
  }

  companion object {
    private val DEFAULT_TITLE_COLOR = Color.CYAN
    private val DEFAULT_TITLE = "Title"
    private val DEFAULT_SHOW_TITLE = true
    private val DEFAULT_TITLE_SIZE = 25f
    private val DEFAULT_BACKGROUND_COLOR = Color.WHITE
    private val DEFAULT_VIEW_SIZE = 96
  }

  @JvmField @field:[State] var titleColor: Int = DEFAULT_TITLE_COLOR
  @JvmField @field:[State] protected var backgroundColor: Int = DEFAULT_BACKGROUND_COLOR
  @JvmField @field:[State] var text: String = DEFAULT_TITLE
  @JvmField @field:[State] var titleSize = DEFAULT_TITLE_SIZE
  @JvmField @field:[State] var showTitle: Boolean = DEFAULT_SHOW_TITLE
  @JvmField @field:[State] var viewSize: Int = 0

  private lateinit var titleTextPaint: TextPaint
  private lateinit var backgroundPaint: Paint

  @State lateinit var innerRectF: RectF

  /**
   * Sets the view's title string attribute value.
   *
   * @param title The example string attribute value to use.
   */
  var titleText: String
    get() = this.text
    set(title) {
      this.text = title
      invalidate()
    }

  override fun onSaveInstanceState(): Parcelable? {
    return Icepick.saveInstanceState(this, super.onSaveInstanceState())
  }

  override fun hasOverlappingRendering(): Boolean {
    return false
  }

  override fun onDraw(canvas: Canvas) {
    innerRectF.set(0f, 0f, viewSize.toFloat(), viewSize.toFloat())

    val centerX = innerRectF.centerX()
    val centerY = innerRectF.centerY()

    canvas.drawCircle(centerX, centerY, (viewSize / 2).toFloat(), backgroundPaint)

    val xPos = centerX.toInt()
    val yPos = (centerY - (titleTextPaint.descent() + titleTextPaint.ascent()) / 2).toInt()

    if (showTitle) {
      canvas.drawText(this.text, xPos.toFloat(), yPos.toFloat(), titleTextPaint)
    }
  }

  override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
    super.onMeasure(widthMeasureSpec, heightMeasureSpec)

    val width = View.resolveSize(DEFAULT_VIEW_SIZE, widthMeasureSpec)
    val height = View.resolveSize(DEFAULT_VIEW_SIZE, heightMeasureSpec)
    viewSize = Math.min(width, height)

    setMeasuredDimension(width, height)
  }

  override fun onRestoreInstanceState(state: Parcelable) {
    super.onRestoreInstanceState(Icepick.restoreInstanceState(this, state))
  }

  fun getBackgroundColor(): Int {
    return backgroundColor
  }

  /**
   * Sets the view's background color attribute value.
   *
   * @param backgroundColor The background color attribute value to use.
   */
  override fun setBackgroundColor(backgroundColor: Int) {
    this.backgroundColor = backgroundColor
    invalidatePaints()
  }

  /**
   * Sets whether the view's title string will be shown.
   *
   * @param flag The boolean value.
   */
  fun setShowTitle(flag: Boolean) {
    this.showTitle = flag
    invalidate()
  }

  private fun init(attrs: AttributeSet?, defStyle: Int) {
    val a = context.obtainStyledAttributes(
        attrs, R.styleable.CircleView, defStyle, 0)

    if (a.hasValue(R.styleable.CircleView_titleText)) {
      this.text = a.getString(R.styleable.CircleView_titleText)
    }

    titleColor = a.getColor(R.styleable.CircleView_titleColor, DEFAULT_TITLE_COLOR)
    backgroundColor = a.getColor(R.styleable.CircleView_backgroundColorValue,
        DEFAULT_BACKGROUND_COLOR)

    titleSize = a.getDimension(R.styleable.CircleView_titleSize, DEFAULT_TITLE_SIZE)

    a.recycle()

    titleTextPaint = TextPaint()
    backgroundPaint = Paint()
    innerRectF = RectF()

    //Title TextPaint
    titleTextPaint.flags = Paint.ANTI_ALIAS_FLAG
    titleTextPaint.isAntiAlias = true
    titleTextPaint.typeface = Typeface.defaultFromStyle(Typeface.BOLD)
    titleTextPaint.textAlign = Paint.Align.CENTER
    titleTextPaint.isLinearText = true
    titleTextPaint.color = titleColor
    titleTextPaint.textSize = titleSize

    //Background Paint
    backgroundPaint.flags = Paint.ANTI_ALIAS_FLAG
    backgroundPaint.style = Paint.Style.FILL_AND_STROKE
    backgroundPaint.color = backgroundColor
  }

  private fun invalidatePaints() {
    backgroundPaint.color = backgroundColor
    invalidate()
  }
}